File: 0226-invert-binary-tree/0226-invert-binary-tree.py
from collections import deque

"""
Match
BFS

Plan
- queue
- append new node 
- popleft, swap left and right nodes if not None
- append new left node and right node to the queue
- repeat until queue is empty 

Evaluate
Time: O(n)
Space: O(n)
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        q = deque()
        q.append(root)
        while len(q) > 0:
            node = q.popleft()
            if node: 
                temp = node.left
                node.left = node.right
                node.right = temp

                q.append(node.left)
                q.append(node.right)
        return root

File: 0226-invert-binary-tree/0226-invert-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

//base case: if root == nullptr, do nothing;


// recursive step
// invert left subtree, invert the right subtree
// switch left and right child nodes for current node

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(!root) {
            return nullptr;
        }
        
        TreeNode * temp = root->left;
        root->left = root->right;
        root->right = temp;
        
        invertTree(root->left);
        invertTree(root->right);
        
        return root;
    }
};